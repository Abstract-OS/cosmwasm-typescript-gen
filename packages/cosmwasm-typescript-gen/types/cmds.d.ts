export namespace Commands {
    export { _boilerplate as boilerplate };
    export { _generate as generate };
    export { _recoil as recoil };
}
import _boilerplate from "./commands/boilerplate";
import _generate from "./commands/generate";
import _recoil from "./commands/recoil";
