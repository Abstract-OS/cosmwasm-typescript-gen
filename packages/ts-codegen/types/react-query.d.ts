declare const _default: (name: string, schemas: any[], outPath: string, options?: { optionalClient?: boolean }) => Promise<void>;
export default _default;
