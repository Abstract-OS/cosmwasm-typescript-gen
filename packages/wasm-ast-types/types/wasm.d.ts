import * as t from '@babel/types';
interface QueryMsg {
    $schema: string;
    title: "QueryMsg";
    oneOf?: any;
    allOf?: any;
    anyOf?: any;
}
interface ExecuteMsg {
    $schema: string;
    title: "ExecuteMsg" | "ExecuteMsg_for_Empty";
    oneOf?: any;
    allOf?: any;
    anyOf?: any;
}
export declare const getMessageProperties: (msg: QueryMsg | ExecuteMsg) => any;
export declare const createWasmQueryMethod: (jsonschema: any) => t.ClassProperty;
export declare const createQueryClass: (className: string, implementsClassName: string, queryMsg: QueryMsg) => t.ExportNamedDeclaration;
export declare const createWasmExecMethod: (jsonschema: any) => t.ClassProperty;
export declare const createExecuteClass: (className: string, implementsClassName: string, extendsClassName: string, execMsg: ExecuteMsg) => t.ExportNamedDeclaration;
export declare const createExecuteInterface: (className: string, extendsClassName: string | null, execMsg: ExecuteMsg) => t.ExportNamedDeclaration;
export declare const propertySignature: (name: string, typeAnnotation: t.TSTypeAnnotation, optional?: boolean) => {
    type: string;
    key: t.Identifier;
    typeAnnotation: t.TSTypeAnnotation;
    optional: boolean;
};
export declare const createTypedObjectParams: (jsonschema: any, camelize?: boolean) => t.ObjectPattern;
export declare const createPropertyFunctionWithObjectParams: (methodName: string, responseType: string, jsonschema: any) => t.TSPropertySignature;
export declare const createQueryInterface: (className: string, queryMsg: QueryMsg) => t.ExportNamedDeclaration;
export declare const createTypeOrInterface: (Type: string, jsonschema: any) => t.ExportNamedDeclaration;
export declare const createTypeInterface: (jsonschema: any) => t.ExportNamedDeclaration;
export {};
