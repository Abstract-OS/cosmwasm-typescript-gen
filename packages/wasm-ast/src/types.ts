export declare type fieldType = 'Long' | 'Coin' | 'Duration' | 'Height' | string;

export interface Field {
    name: string;
    type: fieldType;
    node: any;
};
export interface Interface {
    name: string;
    fields: Field[]
};
