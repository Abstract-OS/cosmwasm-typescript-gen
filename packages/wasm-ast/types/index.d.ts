export * from './types';
export * from './amino';
export * from './messages';
export * from './utils';
