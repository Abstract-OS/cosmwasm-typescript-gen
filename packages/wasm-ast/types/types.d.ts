export declare type fieldType = 'Long' | 'Coin' | 'Duration' | 'Height' | string;
export interface Field {
    name: string;
    type: fieldType;
    node: any;
}
export interface Interface {
    name: string;
    fields: Field[];
}
export interface MessageSchema extends Interface {
    typeUrl: string;
    name: string;
    fields: Field[];
}
export interface EnumConverter {
    funcName: string;
    convertType: 'to' | 'from';
    type: string;
}
export interface Enum {
    filename?: string;
    name: string;
    node: any;
    from?: EnumConverter;
    to?: EnumConverter;
}
export interface Mutation {
    typeUrl: string;
    TypeName: string;
    methodName: string;
}
