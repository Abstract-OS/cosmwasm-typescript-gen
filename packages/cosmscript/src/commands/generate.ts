import { prompt } from '../prompt';
import cosmscript from '../index';
import { sync as glob } from 'glob';
import { readFileSync } from 'fs';

export default async (argv) => {

    const questions = [
        {
            _: true,
            type: 'path',
            name: 'jsonSchemaPath',
            message: 'which directory contains the the Rust contracts?',
            default: './rust'
        },
        {
            _: true,
            type: 'path',
            name: 'outPath',
            message: 'where is the output directory?',
            default: './src'
        },
        {
            _: true,
            type: 'string',
            name: 'name',
            message: 'contract name?'
        }
    ];

    const { jsonSchemaPath, outPath, name } = await prompt(questions, argv);

    const files = glob(jsonSchemaPath + '/**/*.json');
    const schemas = files.map(file => JSON.parse(readFileSync(file, 'utf-8')));
    cosmscript(name, schemas, outPath);
};