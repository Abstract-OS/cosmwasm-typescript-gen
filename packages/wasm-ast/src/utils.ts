import * as t from '@babel/types';
import { snake } from "case";
import { Field } from './types';

export const recursiveNamespace = (names, moduleBlockBody) => {
    if (!names || !names.length) return moduleBlockBody;
    const name = names.pop();
    const body = [
        t.exportNamedDeclaration(
            t.tsModuleDeclaration(
                t.identifier(name),
                t.tsModuleBlock(recursiveNamespace(names, moduleBlockBody))
            )
        )
    ];
    return body;
};

export const arrayTypeNDimensions = (body, n) => {
    if (!n) return t.tsArrayType(body);
    return t.tsArrayType(
        arrayTypeNDimensions(body, n - 1)
    );
};

export const FieldTypeAsts = {
    string: () => {
        return t.tsStringKeyword()
    },
    array: (type) => {
        return t.tsArrayType(FieldTypeAsts[type]())
    },
    Duration: () => {
        return t.tsTypeReference(t.identifier('Duration'))
    },
    Height: () => {
        return t.tsTypeReference(t.identifier('Height'))
    },
    Coin: () => {
        return t.tsTypeReference(t.identifier('Coin'))
    },
    Long: () => {
        return t.tsTypeReference(t.identifier('Long'))
    }
};

export const shorthandProperty = (prop: string) => {
    return t.objectProperty(t.identifier(prop), t.identifier(prop), false, true);
};

export const importStmt = (names: string[], path: string) => {
    return t.importDeclaration(
        names.map(name => t.importSpecifier(t.identifier(name), t.identifier(name))),
        t.stringLiteral(path));
};

export const importAminoMsg = () => {
    return importStmt(['AminoMsg'], '@cosmjs/amino');
};

export const getFieldDimensionality = (field: Field) => {
    let typeName = field.type;
    const isArray = typeName.endsWith('[]');
    let dimensions = 0;
    if (isArray) {
        dimensions = typeName.match(/\[\]/g).length - 1;
        typeName = typeName.replace(/\[\]/g, '');
    }
    return {
        typeName,
        dimensions,
        isArray
    };
}

export const memberExpressionOrIdentifier = (names) => {
    if (names.length === 1) {
        return t.identifier(names[0])
    }
    if (names.length === 2) {
        const [b, a] = names;
        return t.memberExpression(
            t.identifier(a),
            t.identifier(b)
        );
    }
    const [name, ...rest] = names;

    return t.memberExpression(
        memberExpressionOrIdentifier(rest),
        t.identifier(name)
    )
};

export const memberExpressionOrIdentifierSnake = (names) => {
    if (names.length === 1) {
        return t.identifier(snake(names[0]))
    }
    if (names.length === 2) {
        const [b, a] = names;
        return t.memberExpression(
            t.identifier(snake(a)),
            t.identifier(snake(b))
        );
    }
    const [name, ...rest] = names;

    return t.memberExpression(
        memberExpressionOrIdentifierSnake(rest),
        t.identifier(snake(name))
    )
};
