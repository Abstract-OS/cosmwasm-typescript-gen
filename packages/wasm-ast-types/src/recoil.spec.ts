import { importStmt } from './utils'
import generate from '@babel/generator';
import * as t from '@babel/types';

import query_msg from './__fixtures__/schema/query_msg.json';

import {
    createRecoilSelectors,
} from './recoil';

const expectCode = (ast) => {
    expect(
        generate(ast).code
    ).toMatchSnapshot();
}

const printCode = (ast) => {
    console.log(
        generate(ast).code
    );
}

it('query interfaces', () => {
    expectCode(createRecoilSelectors(
        'SG721',
        query_msg
    ))
});
