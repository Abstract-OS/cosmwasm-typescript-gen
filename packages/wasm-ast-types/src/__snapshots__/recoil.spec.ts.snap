// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query interfaces 1`] = `
"export const governanceModulesSelector = selectorFamily<GovernanceModulesResponse | undefined, Parameters<QueryClient[\\"governanceModules\\"]>[0] & {
  contractAddress: string;
}>({
  key: \\"cwGovernaneceGovernanceModules\\",
  get: ({
    contractAddress,
    ...params
  }) => ({
    get
  }) => {
    const client = get(queryClient(contractAddress));
    if (!client) return;
    return await client.governanceModules(params);
  }
});"
`;
